mrc-ide/packit
# mrc-4723 Check user org membership on browser login, and security refactor

This branch implements three separate tickets which, after some investigation, it turned out to make sense to implement together. These are:
- mrc-4702: Check user org membership on browser login (not just API login)
- mrc-4682: Use third party Github API package (kohsuke)
- mrc-4723: Refactor auth to re-use more code between API and browser paths

Essentially, I did an initial implementation of mrc-4702 which used built-in Spring Security OAuth2 classes to access the Github API using the token obtained from the OAuth2 redirected browser login. However when I looked into re-using this approach for Packit API login, this proved infeasible as the Spring Security machinery is quite intractable to make a partial use of (providing the PAT token via our own API rather than obtained during the OAuth2 handshake). So instead, I've refactored so that both paths now call the same client code, which makes use of the third party Github API package, replacing the previous 'roll your own' client. 

Key changes:
- `GithubClient` reimplemented as `GithubUserClient`, largely replicating code from OrderlyWeb. Authenticates with given token, and returns a user principal as well as checking org membership (checking team membership to follow in future ticket). 
- `GithubUserLoginService` renamed to `GithubAPILoginService` - this uses the new `GithubUserClient` and is called from `LoginController` to execute the API login path. 
- `OAuth2UserService` updated to also use `GithubUserClient` to check org membership
- `OAuth2FailureHandler` added, and included in `WebSecurityConfig`. This deals with exceptions thrown during browser login (including org membership check failure). Like the success handler, this redirects to `/redirect` in the front end, but passes an `error` parameter rather than `token`. 
- `BrowserRedirect` pulled out to share some redirect code between the success handler and failure handler.
- The failure handler also uses `PackitExceptionHandler` to deal with PackitAuthenticationExceptions - the handler looks up error messages from the resource bundle. The API login path is already using the exception handler. We have a new `PackitAuthenticationException` type as the failure handler only deals with exceptions inherited from `AuthenticationException`.
- Front end changes to allow the `Redirect` component to deal with redirections with error parameter rather than token parameter. In this case, commit the error and redirect to the login page, which will display the error. 


To test:
Make yourself an OAuth app on Github and set the client id and secret in `/api/app/src/main/resources/application.properties`
Success case: Make sure auth.githubAPIOrgs in  `/api/app/src/main/resources/config.properties` includes an org you are a member of before you run the app
Failure case: Make sure auth.githubAPIOrgs does not include an org you are a member of

Browser login: Run up the app and login from htpp://localhost:3000. You should see a sensible error message in the failure case.
API login:  Use your Github PAT in a post like `curl -v -X POST -H "Content-Type: application/json" -d '{"githubtoken": "[YOUR GITHUB PAT]"}' http://localhost:8080/auth/login/github`.  The PAT needs to have read:org permission. You should get a JWT token and 200 response back in success case, or an error with 401 response in failure case.
